wandb/internal/sender.py

```diff
@@ -258,57 +258,73 @@ def send_request_poll_exit(self, record):
         """This maybe queries the backend for a run and fails if the settings are
         incompatible."""
-        error = None
-        if self._settings.resume:
-            # TODO: This causes a race, we need to make the upsert atomically
-            # only create or update depending on the resume config
-            # we use the runs entity if set, otherwise fallback to users entity
-            entity = run.entity or self._entity
-            logger.info(
-                "checking resume status for %s/%s/%s", entity, run.project, run.run_id
-            )
-            resume_status = self._api.run_resume_status(
-                entity=entity, project_name=run.project, name=run.run_id
+        if not self._settings.resume:
+            return
+
+        # TODO: This causes a race, we need to make the upsert atomically
+        # only create or update depending on the resume config
+        # we use the runs entity if set, otherwise fallback to users entity
+        entity = run.entity or self._entity
+        logger.info(
+            "checking resume status for %s/%s/%s", entity, run.project, run.run_id
+        )
+        resume_status = self._api.run_resume_status(
+            entity=entity, project_name=run.project, name=run.run_id
+        )
+
+        if not resume_status:
+            if self._settings.resume == "must":
+                error = wandb_internal_pb2.ErrorInfo()
+                error.code = wandb_internal_pb2.ErrorInfo.ErrorCode.INVALID
+                error.message = (
+                    "resume='must' but run (%s) doesn't exist" % run.run_id
+                )
+                return error
+            return
+
+        #
+        # handle cases where we have resume_status
+        #
+        if self._settings.resume == "never":
+            error = wandb_internal_pb2.ErrorInfo()
+            error.code = wandb_internal_pb2.ErrorInfo.ErrorCode.INVALID
+            error.message = "resume='never' but run (%s) exists" % run.run_id
+            return error
+
+        history = {}
+        events = {}
+        config = {}
+        summary = {}
+        try:
+            history = json.loads(
+                json.loads(resume_status["historyTail"])[-1]
             )
-            if resume_status is None:
-                if self._settings.resume == "must":
-                    error = wandb_internal_pb2.ErrorInfo()
-                    error.code = wandb_internal_pb2.ErrorInfo.ErrorCode.INVALID
-                    error.message = (
-                        "resume='must' but run (%s) doesn't exist" % run.run_id
-                    )
-            else:
-                if self._settings.resume == "never":
-                    error = wandb_internal_pb2.ErrorInfo()
-                    error.code = wandb_internal_pb2.ErrorInfo.ErrorCode.INVALID
-                    error.message = "resume='never' but run (%s) exists" % run.run_id
-                elif self._settings.resume in ("allow", "auto"):
```

Comment:
This was the main problem: this should have also included "must"

I flattened this function to make it easier to read with all the error paths.
also added an error for resume="must" where there was a problem resuming from history tec